### ISSUE TITLE
table indexes not actually getting dropped on clearing edges

### ISSUE THREAD

**Issue by petrelharp** [tskit-dev/tskit#2317]
I've run into a case where I'm getting errors on doing `tables.tree_sequence()` after clearing/rebuilding the edges because the indexes are stale. Here's a MWE, in which I (1) build some valid tables; (2) clear the edges and build a different set of valid tables; (3) try to get the resulting tree sequence.
```
import tskit

do_error = True

t = tskit.TableCollection(sequence_length=10)
t.nodes.add_row(time=0)
t.nodes.add_row(time=0)
t.nodes.add_row(time=1)
t.nodes.add_row(time=2)

t.edges.add_row(left=0, right=5, parent=2, child=0)
t.edges.add_row(left=0, right=5, parent=2, child=1)
t.edges.add_row(left=5, right=10, parent=3, child=0)
t.edges.add_row(left=5, right=10, parent=3, child=1)

# note it would suffice to just do t.build_index() for the next two lines
ts = t.tree_sequence()
t = ts.dump_tables()
print("Previous indexes:")
print(t.indexes)

t.edges.clear()
print("After clearing:")
print(t.indexes)

t.edges.add_row(left=0, right=5, parent=2, child=0)
t.edges.add_row(left=0, right=8, parent=2, child=1)
t.edges.add_row(left=5, right=10, parent=3, child=0)
t.edges.add_row(left=5, right=8, parent=3, child=2)

if do_error:
    print("Not building indexes.")
else:
    print("Rebuilding indexes.")
    t.build_index()

print("New indexes:")
print(t.indexes)

t.tree_sequence()
```
which produces
```
Previous indexes:
TableCollectionIndexes(edge_insertion_order=array([0, 1, 2, 3], dtype=int32), edge_removal_order=array([1, 0, 3, 2], dtype=int32))
After clearing:
TableCollectionIndexes(edge_insertion_order=None, edge_removal_order=None)
Not building indexes.
New indexes:
TableCollectionIndexes(edge_insertion_order=array([0, 1, 2, 3], dtype=int32), edge_removal_order=array([1, 0, 3, 2], dtype=int32))
::: (5.000000) 0 - 2 not 3
Traceback (most recent call last):
  File "/home/peter/debug/tskit/build_index/test.py", line 40, in <module>
    t.tree_sequence()
  File "/home/peter/.local/lib/python3.9/site-packages/tskit-0.4.1-py3.9-linux-x86_64.egg/tskit/tables.py", line 3267, in tree_sequence
    return tskit.TreeSequence.load_tables(self)
  File "/home/peter/.local/lib/python3.9/site-packages/tskit-0.4.1-py3.9-linux-x86_64.egg/tskit/trees.py", line 3730, in load_tables
    ts.load_tables(tables._ll_tables, build_indexes=build_indexes)
_tskit.LibraryError: Bad edges: contradictory children for a given parent over an interval. (TSK_ERR_BAD_EDGES_CONTRADICTORY_CHILDREN)
```

However, changing `do_error` to False doesn't throw an error and tells us what the indexes ought to have been for the second tree:
```
Previous indexes:
TableCollectionIndexes(edge_insertion_order=array([0, 1, 2, 3], dtype=int32), edge_removal_order=array([1, 0, 3, 2], dtype=int32))
After clearing:
TableCollectionIndexes(edge_insertion_order=None, edge_removal_order=None)
Rebuilding indexes.
New indexes:
TableCollectionIndexes(edge_insertion_order=array([0, 1, 2, 3], dtype=int32), edge_removal_order=array([0, 3, 1, 2], dtype=int32))
```

In short - even though printing out the indexes tells us they've been cleared, somehow they haven't been invalidated, and so doing `tables.tree_sequence()` tries to use them.

**Comment by jeromekelleher** [comment]
This is a tricky one, and I'm not sure there's an easy solution. The problem is that ``tables.edges.clear()`` doesn't alter the indexes, because the edge table doesn't contain a reference to the edge table. The behaviour above comes from the definition of ``has_index`` [here](https://github.com/tskit-dev/tskit/blob/f4b8dcb8d7873fdd4b3b94b2e837f9d181ac0ae8/c/tskit/tables.c#L10661) which just checks that the index arrays are non-null and the number of rows is what is expected.

So, you hit this bug because you happened to put back exactly the same number of edges after clearing.

There would be quite a bit of work involved in doing a fully guaranteed-to-be-coherent index in the face of all possible table updates, so we decided to settle for something that was guaranteed to be safe. We can definitely think about making the coherency guarantees stronger, but I wonder if it's worth it.

**Comment by benjeffery** [comment]
I thought about this when I was changing the indexing flags recently. I guess `has_index` could be made to be more of a `check_index_integrity` but that would mean going over all the edges which wouldn't be a quick check like the current implementation.
The other option would be to have the indexes owned by the edge table, but that's not something I would want to do post 1.0 without major motivation.

**Comment by jeromekelleher** [comment]
I think we need ``has_index`` to be O(1), as this is checked in a fair few places.

We could keep an optional reference to the ``table_collection`` in the edge_table (and others) and clear the index when stuff happens. It gets complicated quickly though - do you check/clear the index for every add_row operation?

**Comment by petrelharp** [comment]
Hm, this *is* complicated. Let's see - I think that the common "dump tables, edit them, make tree sequence again" shouldn't need to have you remember to build indexes or else you get wrong answers. Or is this unreasonable? It seems like at least .clear() should invalidate them - but, I don't think we need to check/clear the index on every add_row, because if the indexes aren't the right length that can be noticed when anything tries to use them.

**Comment by benjeffery** [comment]
I think we could improve the error message that @petrelharp saw to add the suggestion that it can be due to an inconsistent index. I'm reluctant to make the change to `clear` as described above.

**Comment by petrelharp** [comment]
Just brainstorming here, but: `TableCollection.tree_sequence()` could have a try... except clause that if it failed checked if the indexes needed to be rebuilt?

When I filed this it felt bug-ish, but perhaps it is not a bug? (also I don't know why github is telling me I added the bug label 1 minute ago; in fact I added it when I made the issue, but maybe shoudln't have)

**Comment by petrelharp** [comment]
It doesn't seem to me that clearing the indexes on edges.clear would be breaking the contract - unless the docs say explicitly that *doesn't* happen? I guess I can imagine cases where someone would be correctly using the fact that edges.clear doesn't drop indexes, but (a) those seem pretty rare and (b) bugs due to incorrectly using that fact seem much more common.

**Comment by benjeffery** [comment]
I don't think this counts as a bug, maybe "unhelpful behaviour"? I suggest that we we improve the docs on `clear` along with the error message.

**Comment by petrelharp** [comment]
I removed "bug", but I'm pretty sure I could construct an example where by doing this you don't get an error, just the wrong trees. Although it'd be rare, I'd say it counts as "buggy behaviour", so I think we'd be justified in changing it (although finding a lower-resistance path would be great). Changing the error message would probably work just fine, though.

A different thing that could help this not happen is if `TreeSequence.dump_tables()` gave table copies without the indexes?

**Comment by jeromekelleher** [comment]
> but I'm pretty sure I could construct an example where by doing this you don't get an error, just the wrong trees.

I think it'd be worth going through this exercise to verify, as it is a significant amount of work to implement. I'm less worried about this as you do have to put back in exactly the same number of edges, which already makes it a rare occurrence in my book.

**Comment by petrelharp** [comment]
Well, here's a simple example:
```
import tskit

t = tskit.TableCollection(sequence_length=5)
t.nodes.add_row(time=0, flags=1)
t.nodes.add_row(time=0, flags=1)
t.nodes.add_row(time=1)

t.edges.add_row(left=0, right=5, parent=2, child=0)  # 1
t.edges.add_row(left=0, right=5, parent=2, child=1)  # 0

t.build_index()
print("Previous indexes:")
print(t.indexes)

ts1 = t.tree_sequence()

t.edges.clear()

t.edges.add_row(left=0, right=1, parent=2, child=0)
t.edges.add_row(left=0, right=5, parent=2, child=1)

# WRONG
ts2 = t.tree_sequence()

t.build_index()

# RIGHT
ts3 = t.tree_sequence()

print("Wrong tree sequence:")
for t in ts2.trees():
    print("-- tree on interval:", t.interval)
    print(t.draw_text())


print("Right tree sequence:")
for t in ts3.trees():
    print("-- tree on interval:", t.interval)
    print(t.draw_text())
```
This produces:
```
Wrong tree sequence:
-- tree on interval: Interval(left=0.0, right=5.0)
 2 
┏┻┓
0 1

Right tree sequence:
-- tree on interval: Interval(left=0.0, right=1.0)
 2 
┏┻┓
0 1

-- tree on interval: Interval(left=1.0, right=5.0)
  2
  ┃
0 1

```

However, what I'm exploiting here to get a bad tree sequence can be caught; I'll put in a PR with some extra error checking.
